First Version of Intel PCT Getting Started (#37)
intel/optimization-zone: hardware/priority_core_turbo/README.md

# Enabling Priority Core Turbo (PCT) for GPU Performance

## Overview

> **Intel® Priority Core Turbo (PCT)** is part of **Intel® Speed Select Technology – Turbo Frequency (SST-TF)**.

Reviewer: If you have any public links to Intel Speed Select Technology, please add it here.

Author: good idea. added links for PCT, SST, and SST-TF. thanks

It allows a subset of CPU cores to operate at **higher turbo frequencies**, while remaining cores run closer to base frequency.

This is particularly effective for **GPU-accelerated AI inference**, where a small number of CPU threads handle
**latency-critical, mostly serial tasks** such as tokenization, scheduling, and feeding GPUs.

> Running these threads on **High-Priority (HP) cores** improves GPU utilization, TTFT, and tail latency.

Reviewer 2: Please include the expanded version of TTFT. It should be time-to-first-token, right?

Author: yes. added it accordingly. thanks

Validated platforms:

- **Intel® Xeon® 6776P**

## How PCT Works
<details>
<summary> PCT Details </summary>
  
PCT relies on **two Intel Speed Select features**:

- **SST-TF (Turbo Frequency)**
  Defines the high-priority turbo buckets and the number of physical cores that can use each bucket.

- **SST-CP (Core Power / CLOS)**
  Assigns CPUs to **Classes of Service (CLOS)**.
  CPUs assigned to **CLOS0** are treated as **High-Priority** by PCT.

> **Important:** PCT is only effective when CPUs are explicitly assigned to **CLOS0** and Core Power / CLOS is enabled.

### PCT bucket-count interpretation

> `intel-speed-select turbo-freq info -l <level>` may print the same `bucket-0`,

Reviewer: Can you share how user can download or get the intel-speed-select tool ? I dont think it is covered prior to this. 
And show a sample output of info -l <level> ?

Reviewer: I see that  intel-speed-select  is part of the Dockerfile. We should probably have this section after the build section and after tool is introduced.

Reviewer 2: I think in general it would be good to have a section preceding `PCT bucket-count interpretation` that discusses the different software projects that are required/available for PCT, where to get them, and how to install them (even if it's just a reference to scripts used later in the document).

Author: @rsiyer-intel  feel better to explain PCT first, and then help on env setup later. added a refer link to build section in the beginning of the seciton.

Author: @adgubrud  we interpretate for users and then set it according to numa topology in pct_map_and_set_clos.sh  there are just some background knowledge. users don't need to do anything themselves for that interpretation. hope it address your quesiton

Reviewer 2: @louie-tsai I think I understand the SW requirements a bit better now and my previous comment doesn't apply anymore. 

What I'm trying to get at is that a user will likely be coming into this article with no understanding of how to achieve the steps you're outlining. If you mention in the overview that this guide provides a Docker container and supporting scripts without any extra software dependencies, it will be helpful to provide that context.

Author: addressed it accordingly

> `bucket-1`, and `bucket-2` SST-TF table under multiple `powerdomain-*` anchors.

Reviewer 2: Please include a quick description of what a powerdomain anchor is

Author: added it accordingly

For PCT **capacity**, this flow counts `bucket-0` **once per package/socket**:

```text
bucket-0 high-priority-cores-count:8 @ 4600 MHz
=> 8 PCT physical cores per package/socket
```

On a two-socket Intel® Xeon® 6776P system with Hyper-Threading enabled:

```text
2 packages × 8 physical PCT cores/package = 16 physical PCT cores total
16 physical PCT cores × 2 threads/core    = 32 logical PCT CPUs total
```

### Capacity versus placement

There are two different concepts:

| Concept | Correct model |
| --- | --- |
| **PCT capacity** | Count `bucket-0` once per package/socket |
| **HP CPU placement** | Dispatch the package-level PCT core budget across the package's PCT reporting powerdomain anchors |

For Intel® Xeon® 6776P system, `bucket-0` reports:

> ```text

Reviewer 2: Please include the command that produces this output

Author: it is output from check_status script. added some explanation there.

Reviewer 2: Please see my alternate suggestion above

Author: can you elaborate more?

PCT_CORES_PER_PACKAGE=8
PCT_ACTIVE_PACKAGES=2
PCT_TOTAL_PHYSICAL_CORES=16
THREADS_PER_CORE=2
PCT_TOTAL_LOGICAL_CPUS=32
```

> But the `turbo-freq` output shows two reporting anchors per package:

Reviewer: Will be good to specify that this is a 2S 64c per socket - Intel Xeon 6776P system.

Author: added it accordingly

> ```text

Reviewer 2: Please include the command that produces this output

Author: it is parsed results. no direct command to get it. we handle it all in our script.

package 0: anchor cpu0,  anchor cpu32
package 1: anchor cpu64, anchor cpu96
```

Therefore, the set script dispatches the **8 physical PCT cores per package**
across the package's two reporting anchors:

```text
package 0: 4 physical cores from cpu0  + 4 physical cores from cpu32
package 1: 4 physical cores from cpu64 + 4 physical cores from cpu96
```

With Hyper-Threading included, this becomes:

```text
0-3,32-35,64-67,96-99,128-131,160-163,192-195,224-227
```

> This is the default strict bucket-0 PCT placement used by the updated set script.

Reviewer 2: Is the updated set script "set_kernel_env.sh"? Please link to whichever script you're mentioning.

Author: this sentence is not required. removed it accordingly.

</details>

## 1. Build the Environment

<details>
<summary> Build Details </summary>
  
Export the kernel build variables first:

```bash

> source ./set_kernel_env.sh

Reviewer: Can you mention minimum kernel requirement in the main README itself ?
echo "WARN: validated GNR PCT flow expects KERNEL_MM=6.8 and KERNEL_TAG=v6.8." >&2

Author: actually, no minimum kernel. 6.8 just default one.  when user runs set_kernel_env.sh, it will update env based on their kernel version.

intel-speed-select-set-1  | Setting selected HP CPUs -> CLOS0
intel-speed-select-set-1  | Applied.
intel-speed-select-set-1  |
intel-speed-select-set-1  | ------------------------------------------------------------
intel-speed-select-set-1  | Verification (concise CPU->CLOS)
intel-speed-select-set-1  | ------------------------------------------------------------
intel-speed-select-set-1  | HP list should be clos:0
intel-speed-select-set-1  | cpu-0 clos:0
intel-speed-select-set-1  | cpu-1 clos:0
intel-speed-select-set-1  | … (showing first 2 lines)
intel-speed-select-set-1  |
intel-speed-select-set-1  | Non-HP list should be clos:2
intel-speed-select-set-1  | cpu-4 clos:2
intel-speed-select-set-1  | cpu-5 clos:2
intel-speed-select-set-1  | … (showing first 2 lines)
intel-speed-select-set-1  |
intel-speed-select-set-1  | Done.
```

</details>

After applying the set flow, run the check flow again.
The check output should show:

```text
clos:0 -> 32 CPUs
clos:0 CPU list: 0-3,32-35,64-67,96-99,128-131,160-163,192-195,224-227
Expected PCT logical CPU budget   : 32
✅ CLOS0 CPU count exactly matches the bucket-0 PCT logical budget.
```

> ## 4. Benchmark CLOS0 CPUs with Host PerfSpect

Reviewer: Please provide link to PerfSpect, users may not be aware of this tool https://github.com/intel/PerfSpect 
When you say Host PerfSpect, please clarify it as "Benchmark CLOS0 CPUs with PerfSpect tool on the host" (there is no separate Host and Container PerfSpect)

Author: Thanks. I changed title accordingly.
I also have how to install perfspect in folded Details section.
Hope the information works.
<img width="1277" height="573" alt="image" src="https://github.com/user-attachments/assets/c3fe5263-e9bb-40ed-9d49-a5e4772c77b1" />

Use Docker only to configure and verify PCT/CLOS. Run PerfSpect on the host so
the frequency benchmark can access host CPU frequency interfaces directly.

### Prerequisites

<details>
<summary> Details </summary>
  
The host benchmark script reads the CPU list generated by the check profile:

```bash
./results/clos0_cpulist.txt
```

Install PerfSpect on the host first:

```bash
mkdir -p "${HOME}/tools"
cd "${HOME}/tools"

wget -qO- https://github.com/intel/PerfSpect/releases/latest/download/perfspect.tgz | tar -xz

sudo ln -sf "${HOME}/tools/perfspect/perfspect" /usr/local/bin/perfspect
```

Confirm it is available:

```bash
which perfspect
perfspect --help | head
```

</details>

### Run the benchmark

> 

Reviewer 2: Please provide a short description of what the benchmark is going to do. It looks like you're having PerfSpect run its speed and frequency benchmarks so just provide that context and a short description of the kind of outputs we're expecting to see in the results. I also see mention of running the memory benchmark - how is PCT going to affect those results?

Author: added it accordingly.

Run the full flow:

```bash
docker compose --progress=plain --profile set up --abort-on-container-exit
docker compose --progress=plain --profile check up --abort-on-container-exit

> ./run_host_perfspect_benchmark.sh

Reviewer: Please make this a link to the script.

Author: can't add link in code snapshot section, so added the link in the sentence above code snapshot section. hope it works

```

<details>
<summary> Details </summary>

Default host benchmark command:

```bash
sudo taskset -c "${CLOS_CPUS}" perfspect benchmark --speed --frequency --no-summary --output <output-dir>
```

Override the PerfSpect benchmark options with `PERFSPECT_ARGS`:

```bash
PERFSPECT_ARGS="--speed --frequency --memory --no-summary" \
./run_host_perfspect_benchmark.sh
```
</details>

### Analyze results

Benchmark output is written under:

```bash
./results/perfspect_host_clos0_<timestamp>/
```

The directory includes:

```text
clos0_cpulist.txt
perfspect_benchmark.log
perfspect/
```
Check Frequency section in HTML file.  
This is the frequency diagram on Xeon 6776P with PCT on. 
<img width="835" height="483" alt="image" src="https://github.com/user-attachments/assets/96f8855c-4b83-4c62-a0dd-fa2408f979fb" />

> This is the expected pattern: small active core counts hold the highest PCT turbo

Reviewer: Please clarify that the frequency reported is the frequency when "X" number of cores are active.  Can you elaborate what the user needs to understand from above image? Should they be looking at 1-8 cores being active? Are you referring to the SSE frequencies ? They should be ignoring SSE(expected) row.

Reviewer: Frequency steps down as more cores are active, because the average of all the cores is taken ? CLOS0 cores will always operate at higher frequency, no matter how many cores are active?

Author: Since we use CPU as head node, we normally don't use AVX512 or AMX on CPUs for GPU workload, so we look into SSE.  more CPU cores are active and frequency will go down. CLOS0 CPU still share the power with other CPUs in the same power domain, so CLOS0 CPU frequency will also go down when CLOS2/3 CPUs start active. You could see frequency when CPU Core count > 8, frequency drops when it starts using CLOS2/3 CPUs.

frequency, and frequency gradually steps down as more physical cores become active.

## 5. Debug / Manual Inspection (Optional)

<details>
<summary> Debug Details </summary>

> 

Reviewer 2: Are there common pitfalls you expect the user to run into?  What will these useful commands help with?

Author: no. it is just an optional section for further debugging purpose. in general, users don't need to run it themselves.

Reviewer 2: Okay - what about a few word description of what info each command will provide? See below

Author: added them accordingly

### PCT bucket-count interpretation

intel-speed-select tool mentioned below is installed inside the docker image in the [environment build seciton](#1-build-the-environment)

`intel-speed-select turbo-freq info -l <level>` may print the same `bucket-0`,
`bucket-1`, and `bucket-2` SST-TF table under multiple `powerdomain-*` anchors.  
A powerdomain anchor is the representative CPU id where a packages's internal power domain starts.   

For PCT **capacity**, this flow counts `bucket-0` **once per package/socket**:

```text
bucket-0 high-priority-cores-count:8 @ 4600 MHz
=> 8 PCT physical cores per package/socket
```

On a two-socket Intel® Xeon® 6776P system with Hyper-Threading enabled:

```text
2 packages × 8 physical PCT cores/package = 16 physical PCT cores total
16 physical PCT cores × 2 threads/core    = 32 logical PCT CPUs total
```

### Capacity versus placement

There are two different concepts:

| Concept | Correct model |
| --- | --- |
| **PCT capacity** | Count `bucket-0` once per package/socket |
| **HP CPU placement** | Dispatch the package-level PCT core budget across the package's PCT reporting powerdomain anchors |

> For Intel® Xeon® 6776P system with 2 sockets and 64 cores per socket, `bucket-0` reports using check_pct_status.sh in [check-pct-status session](#2-check-pct-status):

Reviewer 2: I think it would be helpful to rephrase like this:

```suggestion
The excerpts below are from execution of (check_pct_status.sh](./check_pct_status.sh) on an Intel® Xeon® 6776P system with 2 sockets and 64 cores per socket. Subsequent sections explain the outputs and how to interpret them.
```

docker compose --progress=plain --profile check up --abort-on-container-exit

./run_host_perfspect_benchmark.sh
```

<details>
<summary> Details </summary>

Default host benchmark command:

```bash
sudo taskset -c "${CLOS_CPUS}" perfspect benchmark --speed --frequency --no-summary --output <output-dir>
```

Override the PerfSpect benchmark options with `PERFSPECT_ARGS`:

```bash
PERFSPECT_ARGS="--speed --frequency --memory --no-summary" \
./run_host_perfspect_benchmark.sh
```
</details>

### Analyze results

Benchmark output is written under:

```bash
./results/perfspect_host_clos0_<timestamp>/
```

The directory includes:

```text
clos0_cpulist.txt
perfspect_benchmark.log
perfspect/
```
Check Frequency section in HTML file.  
This is the frequency diagram on Xeon 6776P with PCT on. 
<img width="835" height="483" alt="image" src="https://github.com/user-attachments/assets/96f8855c-4b83-4c62-a0dd-fa2408f979fb" />

This is the expected pattern: small active core counts hold the highest PCT turbo
frequency, and frequency gradually steps down as more physical cores become active.

## 5. Debug / Manual Inspection (Optional)

<details>
<summary> Debug Details </summary>

> 
> This section is useful for **troubleshooting**, **validation**, or **manual experimentation**
> with Intel® Speed Select and PCT behavior.
> 
> Start an interactive shell with the required tools installed:
> 
> ```bash
> docker compose run --rm intel-speed-select-shell
> ```
> 
> Useful commands:
> 
> ```bash
> intel-speed-select --info
> intel-speed-select turbo-freq info -l 1
> intel-speed-select core-power info
> intel-speed-select -c 0 core-power get-assoc
> ```

Reviewer 2: ```suggestion
```bash
intel-speed-select --info # what info will this give?
intel-speed-select turbo-freq info -l 1 # what info will this give?
intel-speed-select core-power info # what info will this give?
intel-speed-select -c 0 core-power get-assoc # what info will this give
```
```

Author: Address it accordingly

**latency-critical, mostly serial tasks** such as tokenization, scheduling, and feeding GPUs.
Running these threads on **High-Priority (HP) cores** improves GPU utilization, Time-to-first-token (TTFT), and tail latency.  

This guide provides a Docker-based workflow and supporting scripts to check PCT support, configure CLOS assignments, and validate the selected high-priority CPU list. The PCT setup and verification steps run from the container, so users do not need to install `intel-speed-select` or other extra host-side dependencies beyond Docker.  

Validated platforms:

- **Intel® Xeon® 6776P**

## How PCT Works
<details>
<summary> PCT Details </summary>
  
PCT relies on **two Intel Speed Select features**:

- **SST-TF (Turbo Frequency)**
  Defines the high-priority turbo buckets and the number of physical cores that can use each bucket.

- **SST-CP (Core Power / CLOS)**
  Assigns CPUs to **Classes of Service (CLOS)**.
  CPUs assigned to **CLOS0** are treated as **High-Priority** by PCT.

> **Important:** PCT is only effective when CPUs are explicitly assigned to **CLOS0** and Core Power / CLOS is enabled.

> ### PCT bucket-count interpretation

Reviewer 2: ```suggestion
### PCT Bucket-Count Interpretation
```

intel-speed-select tool mentioned below is installed inside the docker image in the [environment build seciton](#1-build-the-environment)

`intel-speed-select turbo-freq info -l <level>` may print the same `bucket-0`,
`bucket-1`, and `bucket-2` SST-TF table under multiple `powerdomain-*` anchors.  
A powerdomain anchor is the representative CPU id where a packages's internal power domain starts.   

For PCT **capacity**, this flow counts `bucket-0` **once per package/socket**:

```text
bucket-0 high-priority-cores-count:8 @ 4600 MHz
=> 8 PCT physical cores per package/socket
```

On a two-socket Intel® Xeon® 6776P system with Hyper-Threading enabled:

```text
2 packages × 8 physical PCT cores/package = 16 physical PCT cores total
16 physical PCT cores × 2 threads/core    = 32 logical PCT CPUs total
```

> ### Capacity versus placement

Reviewer 2: ```suggestion
### Capacity Versus Placement
```

There are two different concepts:

| Concept | Correct model |
| --- | --- |
| **PCT capacity** | Count `bucket-0` once per package/socket |
| **HP CPU placement** | Dispatch the package-level PCT core budget across the package's PCT reporting powerdomain anchors |

> The excerpts below are from execution of (check_pct_status.sh](./check_pct_status.sh) on an Intel® Xeon® 6776P system with 2 sockets and 64 cores per socket. Subsequent sections explain the outputs and how to interpret them.

Reviewer 2: ```suggestion
The excerpts below are from execution of [check_pct_status.sh](./check_pct_status.sh) on an Intel® Xeon® 6776P system with 2 sockets and 64 cores per socket. Subsequent sections explain the outputs and how to interpret them.
```
Link not working as-is. This should fix it.

```text
PCT_CORES_PER_PACKAGE=8
PCT_ACTIVE_PACKAGES=2
PCT_TOTAL_PHYSICAL_CORES=16
THREADS_PER_CORE=2
PCT_TOTAL_LOGICAL_CPUS=32
```

But the `turbo-freq` output shows two reporting anchors per package:

```text
package 0: anchor cpu0,  anchor cpu32
package 1: anchor cpu64, anchor cpu96
```

Therefore, the set script dispatches the **8 physical PCT cores per package**
across the package's two reporting anchors:

```text
package 0: 4 physical cores from cpu0  + 4 physical cores from cpu32
package 1: 4 physical cores from cpu64 + 4 physical cores from cpu96
```

With Hyper-Threading included, this becomes:

```text
0-3,32-35,64-67,96-99,128-131,160-163,192-195,224-227
```

</details>

> ## 1. Build the Environment

Reviewer 2: ```suggestion
## 1. Build The Environment
```

PCT Budget Validation for CLOS0
------------------------------------------------------------
CLOS0 CPU count             : 32
PCT bucket                        : bucket-0
PCT reporting anchors             : 4
PCT active packages/sockets       : 2
PCT cores per package/socket      : 8
PCT physical core budget          : 16
PCT max frequency                 : 4600 MHz
Threads per core                  : 2
Expected PCT logical CPU budget   : 32
✅ CLOS0 CPU count exactly matches the bucket-0 PCT logical budget.

------------------------------------------------------------
Summary
------------------------------------------------------------
✅ PCT turbo tables detected
✅ PCT capacity detected: 16 physical HP cores total, 32 logical CPUs with HT=2
   Count model: bucket-0 counted once per package/socket, not once per powerdomain anchor.
✅ Core Power enabled
✅ CLOS enabled
Done.
```

</details>

The check script writes the current target-CLOS CPU list to:

```text
./results/clos0_cpulist.txt
```

For the example above, `clos0_cpulist.txt` contains 32 logical CPUs. With
Hyper-Threading enabled, that corresponds to 16 physical PCT cores.

> ## 3. Set PCT and Assigned HP CPUs

Reviewer 2: ```suggestion
## 3. Set PCT And Assigned HP CPUs
```

This step **activates PCT in practice** by assigning selected HP CPUs to **CLOS0**.

The setup script intentionally **overwrites existing BIOS/runtime CLOS settings**:

1. Enable Core Power / CLOS.
2. Move **all online CPUs → `OTHER_CLOS`**.
3. Move selected HP CPUs → `HP_CLOS`.

This prevents stale BIOS or previous runtime CLOS assignments from leaving unexpected
CPUs in CLOS0.

> ### Set-script behavior

Reviewer 2: ```suggestion
### Set-Script Behavior
```

<details>
<summary> The setup script performs the following actions: </summary>
  
- Detects PCT capacity from `intel-speed-select turbo-freq info -l <TDP_LEVEL>`.
- Counts `bucket-0` once per package/socket.
- Derives `HP_PER_PACKAGE` from `PCT_CORES_PER_PACKAGE` unless overridden.
- Reads the PCT reporting anchors from `PCT_DOMAIN_ANCHORS`.
- Dispatches each package's `HP_PER_PACKAGE` physical-core budget across that package's reporting powerdomain anchors.
- Selects contiguous physical CPUs starting from each reporting anchor CPU.
- Includes Hyper-Threading siblings by default with `INCLUDE_HT=1`.
- Assigns:
    - **Selected HP CPUs → CLOS0** by default
    - **All remaining CPUs → CLOS2** by default

</details>

Export the kernel build variables first:

```bash
source ./set_kernel_env.sh
```

Run the setup:

```bash
docker compose --progress=plain --profile set up --abort-on-container-exit
```

Or test the selection without changing the system:

```bash
DRY_RUN=1 docker compose --progress=plain --profile set up --abort-on-container-exit
```

> ### Example: package capacity dispatched across reporting powerdomain anchors

Reviewer 2: ```suggestion
### Example: Package Capacity Dispatched Across Reporting Powerdomain Anchors
```

intel-speed-select-set-1  | Setting selected HP CPUs -> CLOS0
intel-speed-select-set-1  | Applied.
intel-speed-select-set-1  |
intel-speed-select-set-1  | ------------------------------------------------------------
intel-speed-select-set-1  | Verification (concise CPU->CLOS)
intel-speed-select-set-1  | ------------------------------------------------------------
intel-speed-select-set-1  | HP list should be clos:0
intel-speed-select-set-1  | cpu-0 clos:0
intel-speed-select-set-1  | cpu-1 clos:0
intel-speed-select-set-1  | … (showing first 2 lines)
intel-speed-select-set-1  |
intel-speed-select-set-1  | Non-HP list should be clos:2
intel-speed-select-set-1  | cpu-4 clos:2
intel-speed-select-set-1  | cpu-5 clos:2
intel-speed-select-set-1  | … (showing first 2 lines)
intel-speed-select-set-1  |
intel-speed-select-set-1  | Done.
```

</details>

After applying the set flow, run the check flow again.
The check output should show:

```text
clos:0 -> 32 CPUs
clos:0 CPU list: 0-3,32-35,64-67,96-99,128-131,160-163,192-195,224-227
Expected PCT logical CPU budget   : 32
✅ CLOS0 CPU count exactly matches the bucket-0 PCT logical budget.
```

> ## 4. Benchmark CLOS0 CPUs with PerfSpect tool on the host

Reviewer 2: ```suggestion
## 4. Benchmark CLOS0 CPUs With PerfSpect Tool On The Host
```

Use Docker only to configure and verify PCT/CLOS. Run PerfSpect on the host so
the frequency benchmark can access host CPU frequency interfaces directly.

### Prerequisites

<details>
<summary> Details </summary>
  
The host benchmark script reads the CPU list generated by the check profile:

```bash
./results/clos0_cpulist.txt
```

Install PerfSpect on the host first:

```bash
mkdir -p "${HOME}/tools"
cd "${HOME}/tools"

wget -qO- https://github.com/intel/PerfSpect/releases/latest/download/perfspect.tgz | tar -xz

sudo ln -sf "${HOME}/tools/perfspect/perfspect" /usr/local/bin/perfspect
```

Confirm it is available:

```bash
which perfspect
perfspect --help | head
```

</details>

> ### Run the benchmark

Reviewer 2: ```suggestion
### Run The Benchmark
```

By using PerfSpect benchmark feature, it generates a diagram of CPU frequency among different number of active CPU cores.  
The diagram helps us to understand whether PCT cores can reach the right CPU frequency.  
Run the full flow with [run_host_perfspect_benchmark.sh](run_host_perfspect_benchmark.sh) :

```bash
docker compose --progress=plain --profile set up --abort-on-container-exit
docker compose --progress=plain --profile check up --abort-on-container-exit

./run_host_perfspect_benchmark.sh
```

<details>
<summary> Details </summary>

Default host benchmark command:

```bash
sudo taskset -c "${CLOS_CPUS}" perfspect benchmark --speed --frequency --no-summary --output <output-dir>
```

Override the PerfSpect benchmark options with `PERFSPECT_ARGS`:

```bash
PERFSPECT_ARGS="--speed --frequency --memory --no-summary" \
./run_host_perfspect_benchmark.sh
```
</details>

> ### Analyze results

Reviewer 2: ```suggestion
### Analyze Results
```
